Update the natural function documentation with newest changes (#77)
wso2/docs-bi: en/docs/learn/references/natural-functions.md

# Natural Functions

> In this tutorial, you will learn how to use Natural Functions in Ballerina Integrator, which allow the function to contain instructions in natural language. 

Reviewer: Why the change to start with an uppercase letter?

Author: I thought the term should be consistent in every places

Reviewer: Here we are referring to it as a concept, not as a reference to something in the UI.

Author: Updated, Please check
https://github.com/wso2/docs-bi/pull/77/commits/14ba78fbab8d6975737959db30a8046e06c17ea8

> In this tutorial, you will learn how to use Natural Functions in Ballerina Integrator, which allow the function to contain instructions in natural language. 

Reviewer 2: ```suggestion
In this tutorial, you will learn how to use Natural Functions in Ballerina Integrator, which allows the function to contain instructions in natural language. 
```

Author: Updated with 8343ebaad98ab71d70c5222c943ff96bb5362608

Such a function is evaluated at runtime with a call to a Large Language Model(LLM). The example uses a Natural Function to analyze blog content to suggest a category and rate it based on predefined criteria.

???+ tip "Natural Programming"

>     To learn more about Ballerina Natural Programming library module, see [Natural Programming](https://central.ballerina.io/ballerina/np/latest).

Reviewer: I think we can remove this. Or maybe we can point to the blog.

```suggestion
    To learn more about natural programming and natural functions, see [Natural Language is Code: A hybrid approach with Natural Programming](https://blog.ballerina.io/posts/2025-04-26-introducing-natural-programming/).
```

> ### Step 2: Define Types

Reviewer: L15 Can we use a more relevant name?

Author: `Generate type definitions for natural Function`
Wdyt?

Reviewer: This was for L15 - project name.

Reviewer: Should also be changed as

Enter `ProjectName` as the project name.

Author: Updated

1. Click on the **`Add Artifacts`** button and select **`Type`** in the **`Other Artifacts`** section.
2. Click on **`+ Add Type`** to add a new type.
3. Add the **`Record Name`** as `Blog`. Then click on the **`JSON`** button and paste the following JSON payload. Select **`Is Closed`** and click on the **`Import`** button. Then click the **`Save`** button.
    ```json
    {
        "title": "Tips for Growing a Beautiful Garden",
        "content": "Spring is the perfect time to start your garden. Begin by preparing your soil with organic compost and ensure proper drainage. Choose plants suitable for your climate zone, and remember to water them regularly. Don't forget to mulch to retain moisture and prevent weeds."
    }
    ```
4. Add another type with the **`Record Name`** as `Review` and paste the following JSON payload. Select **`Is Closed`** and click on the **`Import`** button. Then click the **`Save`** button.
    ```json
    {

>         "name": "Gardening",

Reviewer: This should be fixed - category, not name.

1. In the design view, click on the **`Add Artifact`** button.
2. Select **`HTTP Service`** under the **`Integration as API`** category.
3. Select the **`Create and use the default HTTP listener`** option from the **`Listener`** dropdown.
4. Select the **`Design from Scratch`** option as the **`Service Contract`** and use `/blogs` as the base path.
5. Click on the **`Create`** button to create the new service with the specified configurations.

<a href="{{base_path}}/assets/img/natural-functions/service.png"><img src="{{base_path}}/assets/img/natural-functions/service.png" alt="HTTP Service" width="70%"></a>

### Step 4: Add a Natural Function
1. Click on the **`Add Artifact`** button and select **`Natural Function`** under the **`Other Artifacts`** category.
2. Use `reviewBlog` as the function name. Then click the **`Add Parameter`** button to add a parameter for the Natural Function. Add input parameter as `blog` of type `Blog`, return type of `Review`, and click on the **`Create`** button.

<a href="{{base_path}}/assets/img/natural-functions/natural-function.png"><img src="{{base_path}}/assets/img/natural-functions/natural-function.png" alt="Natural Function" width="70%"></a>

> 

Author: I will check

> 

Author: updated

> 3. Click on the **`Edit`** button to add the function logic.

Author: Addded

> 3. Click on the **`Edit`** button to specify the prompt.

Reviewer: Re: your concern about using prompt, can also say

```suggestion
3. Click on the **`Edit`** button to specify the requirement in natural language (i.e., the prompt).
```

1. In the design view, click on the **`Add Artifact`** button.
2. Select **`HTTP Service`** under the **`Integration as API`** category.
3. Select the **`Create and use the default HTTP listener`** option from the **`Listener`** dropdown.

> 4. Select the **`Design from Scratch`** option as the **`Service Contract`** and add the service base path as `/blogs`.

Reviewer: ```suggestion
4. Select the **`Design from Scratch`** option as the **`Service Contract`** and use `/blogs` as the base path.
```

Reviewer: Update wherever applicable. E.g., L51.

5. Click on the **`Create`** button to create the new service with the specified configurations.
<a href="{{base_path}}/assets/img/natural-functions/service.png"><img src="{{base_path}}/assets/img/natural-functions/service.png" alt="HTTP Service" width="70%"></a>

### Step 4: Add a Natural Function
1. Click on the **`Add Artifact`** button and select **`Natural Function`** under the **`Other Artifacts`** category.

> 2. Add the function name as `reviewBlog`. Then click **`Add Parameter`** button to add input parameters for the Natural Function. Add input parameter as `blog` of type `Blog`, return type of `Review` and click on the **`Create`** button.

Reviewer: click **the** \*\*\`Add Parameter`**

Reviewer: input parameter -> parameter

    ```
   
### Step 5: Update the resource method

> 1. The service will have a default resource named `greeting` with the **`GET`** method. Click on three dots appear in front of the `/blogs` service and select **`Edit`** from menu.

Reviewer 2: ```suggestion
1. The service will have a default resource named `greeting` with the **`GET`** method. Click on the three dots that appear in front of the `/blogs` service and select **`Edit`** from the menu.
```

2. Then click the **`Edit`** button in front of `/greeting` resource.
3. Change the resource HTTP method to **`POST`**.
4. Change the resource name as `review`.
5. Add a payload named `blog` to the resource of type `Blog`.
6. Change the 201 response return type to `Review`.
7. Click on the **`Save`** button to update the resource with the specified configurations.
<a href="{{base_path}}/assets/img/natural-functions/update-resource.png"><img src="{{base_path}}/assets/img/natural-functions/update-resource.png" alt="Resource" width="70%"></a>

### Step 6: Implement resource logic
1. Click on the `review` resource to navigate to the resource implementation designer view.

> 2. Hover to the arrow after start and click the ➕ button to add a new action to the resource.

Reviewer: Hover to -> Hover over

3. Select **`Call Natural Function`** from the node panel.
4. Select the `reviewBlog` function from the suggestions.

> 5. Change the **`Variable Name`** to `review` and **`Variable Type`** as `Review`. 

Reviewer: ```suggestion
5. Change the **`Variable Name`** to `review` and **`Variable Type`** to `Review`. 
```

> 6. For the **`Blog`** parameter, add value as `blog` and click on the **`Save`** button.    

Reviewer 2: ```suggestion
6. For the **`Blog`** parameter, add the value as `blog` and click on the **`Save`** button.    
```

<a href="{{base_path}}/assets/img/natural-functions/call-np.gif"><img src="{{base_path}}/assets/img/natural-functions/call-np.gif" alt="Function Call" width="70%"></a>
7. Add a new node after the `reviewBlog` function call and select **`Return`** from the node panel.
8. Select the `review` variable from the dropdown and click **`Save`**.
<a href="{{base_path}}/assets/img/natural-functions/add-return.png"><img src="{{base_path}}/assets/img/natural-functions/add-return.png" alt="Add Return" width="70%"></a>

> 9. The resource implementation is now complete. The function `reviewBlog` is called with the `blog` content as input and the `review` is returned as the response.

Reviewer 2: ```suggestion
9. The resource implementation is now complete. The function `reviewBlog` is called with the `blog` content as input, and the `review` is returned as the response.
```

### Step 7: Configure model for Natural Function

> 1. Press `Ctrl + Shift + P` on Windows and Linux, or `shift + ⌘ + P` on a Mac and type `>Ballerina: Configure default model for Natural Functions (Experimental)` to configure the default model for Natural Functions.   

Reviewer 2: ```suggestion
1. Press `Ctrl + Shift + P` on Windows and Linux, or `Shift + ⌘ + P` on a Mac, and type `>Ballerina: Configure default model for Natural Functions (Experimental)` to configure the default model for Natural Functions.   
```

        "content": "For those who want a 360-degree approach to self-care, with advice for betterment in the workplace, home, gym, and on the go, look no further. The Healthy Maven offers recipes for every type of meal under the sun (salads, sides, soups, and more), DIY tips (you’ll learn how to make your own yoga mat spray), and quick workouts. If you like where all this is going, there’s a supplementary podcast run by blogger Davida with guest wellness experts."
    }
    ```

> 5. The response will be a review of the blog content with the category and rating (The rating value may not always be identical across different executions).

Reviewer: I think the fact that the rating can change is obvious. Technically category can too. So if we are saying something we should say it all.

Author: Will remove that

5. The response will be a review of the blog content with the category and rating.

>     ```json

Reviewer 2: Add empty new line before and after of each code block and image links. check all the places

Author: Updated with https://github.com/wso2/docs-bi/commit/8343ebaad98ab71d70c5222c943ff96bb5362608
